### PR TITLE
resolve #122 100% test coverage

### LINKED ISSUE #122: 100 % test coverage
By getting 100 % test coverage we will get all edge cases covered, so that they won't need to be tested manually and nothing will break.
Only a small amount of code is not tested yet.
A good start could be from testing CRUD pagination search endpoints. (`?query=sometext`)

### PR DESCRIPTION
**Pull request by xiaoxianma**

### CODE REVIEW

File: api/crud.py [hunk 1]
```diff
@@ -261,10 +261,6 @@ async def products_add_related(items: Iterable[models.Product]):
     return items
```
Comment (xiaoxianma) [inline api/crud.py:263]: this may be legacy code. api.views.get_products is used instead.

Comment (MrNaif2018) [reply api/crud.py:263]: Oh indeed, you're right. Thank you

File: api/models.py [hunk 2]
```diff
@@ -270,10 +268,10 @@ async def create(cls, **kwargs):
 
         store_id = kwargs["store_id"]
         kwargs["status"] = "Pending"
-        if not store_id:
+        if not store_id:  # pragma: no cover, crud.create_invoice checked
             raise HTTPException(422, "No store id provided")
         store = await Store.get(store_id)
-        if not store:
+        if not store:  # pragma: no cover, crud.create_invoice checked
```
Comment (MrNaif2018) [inline api/models.py:274]: This line can be triggered if `store_id` doesn't exist in db

File: api/utils.py [hunk 3]
```diff
@@ -302,7 +303,7 @@ async def get_one(self, model_id: int, request: Request):
         except HTTPException:
             if self.get_one_auth:
                 raise
-            user = None
+            user = None  # pragma: no cover
```
Comment (MrNaif2018) [inline api/utils.py:306]: This line is covered when accessing public stores

File: api/utils.py [hunk 4]
```diff
@@ -337,17 +332,12 @@ async def put(
             user: Union[None, ModelView.schemes.User] = Security(self.auth_dependency, scopes=self.scopes["put"]),
         ):  # type: ignore
             item = await self._get_one(model_id, user, True)
-            try:
+            with safe_db_write():
+
```
Comment (MrNaif2018) [inline api/utils.py:336]: Just a question: was this space added by black or not? If not, remove it please

File: tests/test_crud.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import pytest
+
+from api import crud
+
+
```
Comment (MrNaif2018) [inline tests/test_crud.py:5]: You can use 
`pytestmark = pytest.mark.asyncio`
In this file

File: tests/test_utils.py [hunk 6]
```diff
@@ -62,3 +70,42 @@ async def test_get_template(notification_template, async_client, token):
...
+@pytest.mark.asyncio
+async def test_notify_template():
+    template = await utils.get_notify_template(MockStore(), "invoice")
+    assert template.startswith("New order from")
+
+
+@pytest.mark.asyncio
+async def test_product_template(async_client, token):
+    qty = 10
+    resp = await async_client.post(
```
Comment (MrNaif2018) [inline tests/test_utils.py:84]: Why is this post request needed? What does it do in this case?

File: tests/test_utils.py [hunk 7]
```diff
@@ -62,3 +70,42 @@ async def test_get_template(notification_template, async_client, token):
...
+    )
+    assert resp.status_code == 200
+    template = await utils.get_product_template(MockStore(), MockProduct(), qty)
+    assert template == f"Thanks for buying  x {qty}!\nIt'll ship shortly!\n"
+    await async_client.delete(f"/templates/{resp.json()['id']}", headers={"Authorization": f"Bearer {token}"})  # cleanup
+
+
+@pytest.mark.asyncio
+async def test_store_template(async_client, token):
+    resp = await async_client.post(
```
Comment (MrNaif2018) [inline tests/test_utils.py:95]: Same here

File: tests/test_utils.py [hunk 8]
```diff
@@ -62,3 +70,42 @@ async def test_get_template(notification_template, async_client, token):
...
+    template = await utils.get_store_template(MockStore(), [MockProduct()])
+    assert template.startswith("Welcome to our shop")
+    await async_client.delete(f"/templates/{resp.json()['id']}", headers={"Authorization": f"Bearer {token}"})  # cleanup
+
+
+@pytest.mark.parametrize("exist", [True, False])
+def test_run_host(exist):
+    if exist:
+        with tempfile.NamedTemporaryFile() as temp:
+            utils.run_host("echo hello", target_file=temp.name)
```
Comment (MrNaif2018) [inline tests/test_utils.py:108]: You might want to create a temporary file, write to it and check result for test completeness

File: tests/test_views_paginations.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    assert resp.json()["count"] == 2
+
+
+def test_next_prev_url(client: TestClient, token: str):
+    # current url
+    resp = client.get("/users?limit=1", headers={"Authorization": f"Bearer {token}"})
+    next_url = resp.json()["next"]
+    next_endpoint = next_url.split("testserver")[-1]
+    assert next_endpoint == "/users?limit=1&offset=1"
+    # next
```
Comment (MrNaif2018) [inline tests/test_views_paginations.py:21]: Maybe
```suggestion
    # previous
```
?

**Review by MrNaif2018 — CHANGES_REQUESTED** [review]
Thank you very much for your PR! Please take a look at review comments!

**Comment by MrNaif2018** [comment]
Note: there are some unstable lines currently.
See [this test run](https://app.circleci.com/pipelines/github/bitcartcc/bitcart/595/workflows/21ed1798-9216-4e81-bc76-bc50ee3ac7c3/jobs/1604)
It shows coverage 99%, uncovered lines in views.py:
639-640, 674
In some runs they get covered, in some they don't. They seem to be unstable.
It's those lines: https://github.com/xiaoxianma/bitcart/blob/master/api/views.py#L638
And https://github.com/xiaoxianma/bitcart/blob/master/api/views.py#L674
on_disconnect calls. Maybe we should just add pragma: no cover on them?
Probably sometimes it's not called as test run finished before that?

File: api/models.py [hunk 10]
```diff
@@ -101,8 +101,6 @@ def update(self, **kwargs):
     async def apply(self):
         if self.wallets:
             await WalletxStore.delete.where(WalletxStore.store_id == self._instance.id).gino.status()
-        if self.wallets is None:
```
[the remark this answers is not in the document]
Comment (xiaoxianma) [reply api/models.py:104]: Based on schemes.CreateStore, "wallets" is a required field with type List[int]. I can update it to []

File: api/models.py [hunk 11]
```diff
@@ -270,10 +268,10 @@ async def create(cls, **kwargs):
 
         store_id = kwargs["store_id"]
         kwargs["status"] = "Pending"
-        if not store_id:
+        if not store_id:  # pragma: no cover, crud.create_invoice checked
```
[the remark this answers is not in the document]
Comment (xiaoxianma) [reply api/models.py:271]: Yeah, my originally though is removing this because crud.create_invoice already check it

File: api/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrNaif2018) [reply api/models.py:104]: Oh,I missed that. Once again unneeded safeguard (:
Alright then

File: api/models.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrNaif2018) [reply api/models.py:271]: I guess it's fine to remove

File: api/utils.py (hunk 3 above)
Comment (xiaoxianma) [reply api/utils.py:306]: thanks. I will add it

File: api/utils.py (hunk 4 above)
Comment (xiaoxianma) [reply api/utils.py:336]: good catch! maybe it's me

File: tests/test_crud.py (hunk 5 above)
Comment (xiaoxianma) [reply tests/test_crud.py:5]: and then remove @pytest.mark.asyncio?

File: tests/test_utils.py (hunk 6 above)
Comment (xiaoxianma) [reply tests/test_utils.py:84]: good to learn. I thought we need a unique template for each category

File: tests/test_crud.py (hunk 5 above)
Comment (MrNaif2018) [reply tests/test_crud.py:5]: Yes

File: tests/test_utils.py (hunk 8 above)
Comment (xiaoxianma) [reply tests/test_utils.py:108]: good point. I will add a content check. Thanks

File: tests/test_crud.py (hunk 5 above)
Comment (xiaoxianma) [reply tests/test_crud.py:5]: will do

**Review by MrNaif2018 — APPROVED** [review]
Looking good! Thank you very much for your PR! Merging!
